omnilib/control.lua

```diff
@@ -1,408 +1,420 @@
+		recipe.enabled = enabled_override
+		if not recipe.enabled then -- Nothing further to do if we killed it
+			return
+		end
+	end
+	-- Update according to compression research status
+	if not recipe_tree and 
+		force_techs["compression-recipes"] and
+		force_techs["compression-recipes"].researched and 
+		not recipe.category:find("-compressed", nil, true) and 
+		not name:find("concentrated", nil, true) then
+		local recipes = recipe.force.recipes
+		-- Handle both compressed and permuted recipes.
+		local compressed, permuted_compressed = recipes[string.format("%s-compression",name)], recipes[name:gsub("omniperm","compression-omniperm")]
+		if compressed then
+			compressed.enabled = true
+		end
+		if permuted_compressed then
+			permuted_compressed.enabled = true
+		end
+		-- Deal with generator variants
+		for tier=1, math.huge do
+			local compressed_recipe = recipes[string.format("%s-concentrated-grade-%i", name, tier)]
+			if compressed_recipe then
+				compressed_recipe.enabled = true
+			else
+				break
 			end
-			if found_it then return true end
-		else
-			if string.find(str,list[i]) then return true end
 		end
 	end
-	return false
+	if recipe.enabled then
+		update_last_tier(recipe)
+	end
 end
-local initilizing = false
-function omni_update(game)
-	initilizing = true
-	log("running through omnidate")
-  for _, force in pairs(game.forces) do
-    log("researching compressed techs where the standard is done.")
-    for _,tech in pairs(force.technologies) do
-			if not start_with(tech.name,"omnipressed-") and tech.force.technologies["omnipressed-" .. tech.name] then
-				tech.force.technologies["omnipressed-" .. tech.name].researched = tech.researched
-      elseif start_with(tech.name,"omnipressed-") and tech.force.technologies[string.sub(tech.name,13)] then
-        tech.force.technologies[string.sub(tech.name,13)].researched = tech.researched
-      end
-    end
-    
-		if force.technologies["compression-recipes"] and force.technologies["compression-recipes"].researched then
-			log("enabling compressed recipes of researched techs")
-      for _, rec in pairs(force.recipes) do
-				if not string.find(rec.name,"compression") and rec.enabled then
-					local compressCheck = rec.name.."-compression"
-					if string.find(rec.name,"omniperm") then
-						local b=string.find(rec.name,"omniperm")
-            compressCheck=string.sub(rec.name,1,b-1).."compression-"..string.sub(rec.name,b,-1)
-          end
-					if force.recipes[compressCheck] then
-						force.recipes[compressCheck].enabled = true
-						local z=1
-						while force.recipes[rec.name.."-compression-grade-"..z] do
-							force.recipes[rec.name.."-compression-grade-"..z].enabled=true
-							z=z+1
+
+local building_tiers = {
+	compact = "compression-compact-buildings",
+	nanite = "compression-nanite-buildings",
+	quantum = "compression-quantum-buildings",
+	singularity = "compression-singularity-buildings"
+}
+
+local function update_tech(tech)
```

Comment:
something is not working here, can you please check that as a compressed tech unlocks, it also gets the paired one done, and also the other way, it is not functioning on my test map. (It works well for multi-tiered and infinite tech, but is failing on regular tech like angels smelting and omnitraction)

Reply:
Fixed below along with one other possible loop as well.